src/librustc_mir/transform/nll/infer.rs

```diff
@@ -0,0 +1,206 @@
+                        // constraint, the "point" is always the point
+                        // AFTER the action that induced the
+                        // constraint. So report the error on the
+                        // action BEFORE that.
+                        assert!(constraint.point.statement_index > 0);
+                        let p = Location {
+                            block: constraint.point.block,
+                            statement_index: constraint.point.statement_index - 1,
+                        };
+
+                        self.errors.push(InferenceError {
+                            constraint_point: p,
+                            name: sup_def.name,
+                        });
+                    }
+                }
+
+                debug!("    sup (after) : {:?}", sup_def.value);
+                debug!("    changed     : {:?}", changed);
+            }
+            debug!("\n");
+        }
+
+        mem::replace(&mut self.errors, IndexVec::new())
+    }
+}
+
+struct Dfs<'a, 'gcx: 'tcx + 'a, 'tcx: 'a> {
+    infcx: InferCtxt<'a, 'gcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+}
+
+impl<'a, 'gcx: 'tcx, 'tcx: 'a> Dfs<'a, 'gcx, 'tcx> {
+    fn new(infcx: InferCtxt<'a, 'gcx, 'tcx>, mir: &'a Mir<'tcx>) -> Self {
+        Self { infcx, mir }
+    }
+
+    fn copy(
+        &mut self,
+        from_region: &Region,
+        to_region: &mut Region,
+        start_point: Location,
+    ) -> bool {
+        let mut changed = false;
+
+        let mut stack = vec![];
+        let mut visited = FxHashSet();
+
+        stack.push(start_point);
+        while let Some(p) = stack.pop() {
+            debug!("        dfs: p={:?}", p);
+
+            if !from_region.may_contain(p) {
+                debug!("            not in from-region");
+                continue;
+            }
+
+            if !visited.insert(p) {
+                debug!("            already visited");
+                continue;
+            }
+
+            changed |= to_region.add_point(p);
+
+            let block_data = self.mir[p.block];
+            let successor_points = if p.statement_index < block_data.statements.len() {
+                vec![Location {
+                    statement_index: p.statement_index + 1,
+                    ..p
+                }]
+            } else {
+                block_data.terminator()
+                    .successors()
+                    .iter()
+                    .map(|&basic_block| Location {
+                        statement_index: 0,
+                        block: basic_block,
+                    })
+                    .collect::<Vec<_>>()
+            };
+
+            if successor_points.is_empty() {
+                // If we reach the END point in the graph, then copy
+                // over any skolemized end points in the `from_region`
+                // and make sure they are included in the `to_region`.
+                for region_decl in self.infcx.tcx.tables.borrow().free_region_map() {
```

Comment:
@nikomatsakis this is the only part I can't figure out. It seems we lose track of the free regions when we convert them all to region variables. See mod.rs line 49:
```rust
self.infcx.tcx.fold_regions(value, &mut false, |_region, _depth| {
    self.regions.push(Region::default());
    self.infcx.next_region_var(rustc_infer::MiscVariable(DUMMY_SP))
})
```
where `_region` might be `RegionKind::ReFree(FreeRegion)`. Also, I'm not sure if I can even use `free_region_map` like I have here at all. This was how (I think) it worked the last time I tried to get this information (many, many months ago), but it seems to have changed a lot since then. Is there a way to query this information from `tcx` or should we keep track of which regions are free regions in `mod.rs` and use that information here?

Reply:
Yeah, so, I was expecting us to not handle free regions yet. We are going to have to figure out (eventually) the set of free-regions in scope for a given MIR block -- though I do not think we will want to modify that particular code in question. That is, the fact that the old region inferencing code wound up assigning a free region is not of particular interest.

We *will* however want to look at the public "signature" of the function, since *that* is defined by the user. Well, modulo closures, where inference plays a role: doing a better job of handling closures is something we'll have to think about too.

Reply:
TL;DR I think i'd just leave this section as a "TODO" for now